commit/src/domain.rs

```diff
@@ -0,0 +1,364 @@
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+use p3_field::{
+    batch_multiplicative_inverse, cyclic_subgroup_coset_known_order, ExtensionField, Field,
+    TwoAdicField,
+};
+use p3_matrix::dense::RowMajorMatrix;
+use p3_matrix::Matrix;
+use p3_util::{log2_ceil_usize, log2_strict_usize};
+
+use crate::two_adic_decompose::decompose;
+
+pub struct LagrangeSelectors<T> {
+    pub is_first_row: T,
+    pub is_last_row: T,
+    pub is_transition: T,
+    pub inv_zeroifier: T,
+}
+
+pub trait PolynomialDomain: Copy {
+    type Val: Field;
+
+    fn size(&self) -> usize;
+
+    fn next_point<Ext: ExtensionField<Self::Val>>(&self, x: Ext) -> Ext;
+
+    // There are many choices for this, but we must pick a canonical one
+    // for both prover/verifier determinism and LDE caching.
+    fn create_disjoint_domain(&self, min_size: usize) -> Self;
+
+    // decompose evaluations until the domain size is less than max_size
+    // and return the resulting domain
+    fn decompose(
+        &self,
+        evals: RowMajorMatrix<Self::Val>,
+        max_size: usize,
+    ) -> RowMajorMatrix<Self::Val>;
+    fn decomposed_domain(&self, max_size: usize) -> Self;
+    fn decomposed_point<Ext: ExtensionField<Self::Val>>(&self, point: Ext, max_size: usize) -> Ext;
+    fn decomposed_width_factor(&self, max_size: usize) -> usize;
+
+    fn recompose<Ext: ExtensionField<Self::Val>>(
+        &self,
+        evals: &[Ext],
+        max_size: usize,
+        point: Ext,
+    ) -> Vec<Ext>;
+
+    // Unnormalized
+    fn selectors_at_point<Ext: ExtensionField<Self::Val>>(
+        &self,
+        point: Ext,
+    ) -> LagrangeSelectors<Ext>;
+
+    // Unnormalized
+    fn selectors_on_coset(&self, coset: Self) -> LagrangeSelectors<Vec<Self::Val>>;
+}
+
+#[derive(Copy, Clone)]
+pub struct TwoAdicDomain<Val: TwoAdicField> {
```

Comment:
nit: maybe something like `MultiplicativeCoset` would be more precise, to distinguish it from additive groups or two-adic circle groups

Reply:
`TwoAdicMultiplicativeCoset` is ok?